src/Microsoft.AspNet.HttpOverrides/OverrideHeaderMiddleware.cs

```diff
@@ -31,21 +34,27 @@ public Task Invoke(HttpContext context)
                 var xForwardedForHeaderValue = context.Request.Headers.GetCommaSeparatedValues(XForwardedForHeaderName);
                 if (xForwardedForHeaderValue != null && xForwardedForHeaderValue.Count > 0)
                 {
-                    IPAddress originalIPAddress;
-                    if (IPAddress.TryParse(xForwardedForHeaderValue[0], out originalIPAddress))
+                    IPAddress IPFromHeader;
+                    if (IPAddress.TryParse(xForwardedForHeaderValue[0], out IPFromHeader))
                     {
-                        if (context.Connection.RemoteIpAddress != null)
+                        var remoteIPString = context.Connection.RemoteIpAddress?.ToString();
+                        if (!string.IsNullOrEmpty(remoteIPString))
                         {
-                            var ipList = context.Request.Headers.Get(XForwardedForHeaderName);
-                            context.Request.Headers.Set(XForwardedForHeaderName, (ipList + "," + context.Connection.RemoteIpAddress.ToString()));
+                            context.Request.Headers.Set(XOriginalIPName, remoteIPString);
                         }
-                        context.Connection.RemoteIpAddress = originalIPAddress;
+                        context.Connection.RemoteIpAddress = IPFromHeader;
                     }
                 }
             }
 
             if ((_options.ForwardedOptions & ForwardedHeaders.XForwardedHost) != 0)
             {
+                var hostString = context.Request.Host.ToString();
+                if (!string.IsNullOrEmpty(hostString))
+                {
+                    context.Request.Headers.Set(XOriginalHostName, hostString);
```

Comment:
You should only set this if you're going to modify Request.Host. Same for Scheme.